Support IMX8ULP

TangleZ opened the pull request.

Commit:
drivers: imx: ipc: Make MU support imx8ulp

Make mu implementation generic. imx8ulp platform has a new version
of mu, then update mu for preparing to support imx8ulp.

Signed-off-by: Zhang Peng <peng.zhang_8@nxp.com>

Commit:
dai: Add dai ops get_srcid

Modules have resource srcid which is use to config dma, we can't get
the srcid from exist dai ops, then add this to implement.
The function get_srcid gets srcid from modules, then set it to dma.

Signed-off-by: Zhang Peng <peng.zhang_8@nxp.com>

Commit:
sof: Add address convert function on platform header

DSP memory region has mapped to Arm memory region on imx8ulp:
ARM: 0x8e000000~0x8fffffff --- DSP: 0x1a000000~0x1bffffff
Then if we want access the address from ARM, need to convert the
address.

Signed-off-by: Zhang Peng <peng.zhang_8@nxp.com>

Commit:
drivers: imx: edma: Prepare for support imx8ulp

Add address convert for edma can visit the address.
Add srcid to config edma for modules can send dma request to edma.

Signed-off-by: Zhang Peng <peng.zhang_8@nxp.com>

Commit:
drivers: imx: Add new file interrupt-generic.c

Add the file for the platform without irqsteer. Now if the platform
have irqsteer we choose interrupt-irqsteer.c, if not then we choose
interrupt-generic.c

Signed-off-by: Zhang Peng <peng.zhang_8@nxp.com>

Commit:
platform: Add platform imx8ulp

Prepare to support imx8ulp, it has DSP and is samilar with imx8mp,
imx8. The difference is DSP has a mapping to ARM, not connect to a
codec, has new version of mu, and a updated edma on imx8ulp.

Signed-off-by: Zhang Peng <peng.zhang_8@nxp.com>

Commit:
platform: Update Kconfig and CMakeLists for imx8ulp

Update Kconfig and CMakeLists, add imx8ulp in scripts xtensa-build-all.sh

Signed-off-by: Zhang Peng <peng.zhang_8@nxp.com>

Commit:
topology: Add sof-imx8ulp-btsco.m4 for imx8ulp

IMX8ULP pipeline is HOST-->VOLUME-->DAI(sai)-->CODEC(bluetooth).
Use edma copy data.

Signed-off-by: Zhang Peng <peng.zhang_8@nxp.com>

TangleZ commented:
Have some files license and the author is not correct, will update it later.

dbaluta commented on src/include/sof/lib/dai.h:
```diff
@@ -340,7 +342,6 @@ static inline int dai_trigger(struct dai *dai, int cmd, int direction)
 	int ret = dai->drv->ops.trigger(dai, cmd, direction);
 
-
 	return ret;
```
Stray new line here.

dbaluta commented on src/include/sof/lib/dai.h:
```diff
@@ -329,7 +332,6 @@ static inline int dai_set_config(struct dai *dai,
 {
 	int ret = dai->drv->ops.set_config(dai, config);
 
-
```
stray new line here.

dbaluta commented on src/include/sof/lib/dai.h:
```diff
@@ -362,7 +362,6 @@ static inline int dai_pm_context_restore(struct dai *dai)
 	int ret = dai->drv->ops.pm_context_restore(dai);
 
-
 	return ret;
```
same here. Please check this commit. You have multiple unerlated new lines removals. Do not add them in this commit.

dbaluta commented on src/audio/dai.c:
```diff
@@ -816,6 +816,7 @@ static void dai_config_reset(struct comp_dev *dev)
 static int dai_prepare(struct comp_dev *dev)
```
Please explain WHY this patch is needed. Your commit message says WHAT the patch does. While this is also useful the commit message should first explain WHY the commit is needed and then add details about it.

Commit:
drivers: imx: edma: Correct burst_elems size

edma only copy 32bit data before, burst_elems should be adjust
according to src width.

Signed-off-by: Zhang Peng <peng.zhang_8@nxp.com>

dbaluta commented on src/include/sof/drivers/mu.h:
```diff
@@ -13,27 +13,61 @@
+#define IMX_MU_xSR(type, index)        (type & IMX_MU_V2 ? index : 0x20)
+#define IMX_MU_xSR_GIPn(type, x) (type & IMX_MU_V2 ? BIT(x) : BIT(28 + (3 - (x))))
+#define IMX_MU_xSR_RFn(type, x) (type & IMX_MU_V2 ? BIT(x) : BIT(24 + (3 - (x))))
+#define IMX_MU_xSR_TEn(type, x) (type & IMX_MU_V2 ? BIT(x) : BIT(20 + (3 - (x))))
```
Add here type == IMX_MU_V2 instead of type & IMX_MU_V2

dbaluta commented on src/include/sof/drivers/sai.h:
```diff
@@ -247,15 +251,28 @@
+#define SAI_CLOCK_DIV		0x5
+#endif
 #define SAI_TDM_SLOTS		2
 
```
Why is this needed? there is no explanation in the commit message.

dbaluta commented on src/drivers/imx/sai.c:
```diff
@@ -31,6 +31,7 @@ static void sai_start(struct dai *dai, int direction)
 {
 	dai_info(dai, "SAI: sai_start");
```
Explain what is src id? Maybe create a separate commit introducing src id.

dbaluta commented on src/drivers/imx/interrupt-generic.c:
```diff
@@ -0,0 +1,51 @@
+
+/* fa00558c-d653-4851-a03a-b21f125a9524 */
+DECLARE_SOF_UUID("noirq-imx", noirq_imx_uuid, 0xfa00558c, 0xd653, 0x4851,
+		 0xa0, 0x3a, 0xb2, 0x1f, 0x12, 0x5a, 0x95, 0x24);
```
This should be named generic-irq-imx. And thus noirq_imx_uuid should be read as generic_irq_imx_uuid.

Commit:
drivers: imx: sai: Add new sai ops: getsrcid

SAI may has resource srcid according to different platform, we have
no ops to get this resource, then add new ops getsrcid to get.
The srcid is used to config dma then dma can receive dma request from
sai.

Signed-off-by: Zhang Peng <peng.zhang_8@nxp.com>

Commit:
include: drivers: sai: Add special fclk for imx8ulp

We expect fclk is 16KHz on imx8ulp, and now it's 48KHz.According to
bclk = mclk / ((DIV + 1) * 2), set SAI_CLOCK_DIV is 0x17.

Signed-off-by: Zhang Peng <peng.zhang_8@nxp.com>

Commit:
drivers: imx: sai: Config hw channel is 1 for 8ulp

The default hw param channel is 2, this will cause dai_verify_params error
when set channel is 1 on imx8ulp.
Add special config for imx8ulp.

Signed-off-by: Zhang Peng <peng.zhang_8@nxp.com>

Commit:
topology: pipe-volume-playback: Correct PCM_CAPABILITIES

The PCM_CAPABILITIES paramter channel min is fixed with 2, it's not
right when channel is small than 2.

Signed-off-by: Zhang Peng <peng.zhang_8@nxp.com>

dbaluta commented on src/drivers/imx/sai.c:
```diff
@@ -425,6 +430,7 @@ const struct dai_driver sai_driver = {
 		.get_fifo		= sai_get_fifo,
+		.get_srcid		= sai_get_srcid,
 		.get_hw_params		= sai_get_hw_params,
 	},
```
This patch will no compile. Please add the patches in an order that will not break bisection.

dbaluta commented on src/include/sof/drivers/sai.h:
```diff
@@ -239,6 +241,8 @@
+#define SAI_FIFO_WORD_SIZE	16
 #else
 #define SAI_FIFO_WORD_SIZE	64
 #endif
```
add the changes for SAI_OFS and FIFO_WORD_SIZE in a separate patch.

dbaluta commented on src/drivers/imx/sai.c:
```diff
@@ -424,7 +424,11 @@ static int sai_get_hw_params(struct dai *dai,
+#else
 	params->channels = 2;
+#endif
 	params->buffer_fmt = 0;
```
I'm fine with this for now. In the future the number of channels should be read from sai->params.

dbaluta commented on src/include/sof/drivers/sai.h:
```diff
@@ -251,8 +251,14 @@
  */
+#ifndef CONFIG_IMX8ULP
 #define SAI_CLOCK_DIV		0x7
 #define SAI_TDM_SLOTS		2
```
put the condition the other way around #ifdef CONFIG_IMX8ULP. I'm fine with this for now, but in the future we should make

dbaluta commented on src/include/sof/lib/dai.h:
```diff
@@ -417,6 +420,18 @@ static inline int dai_get_fifo(struct dai *dai, int direction,
+}
+
 /**
  * \brief Digital Audio interface Probe
```
This patch should be much earlier in the series so that we avoid breaking the bisection. E.g after applying each patch the code should still compile.

iuliana-prodan replied:
Yes, I agree with Daniel.
Definitely must be added earlier, at least before commit no. 2 - "drivers: imx: sai: Add new sai ops: getsrcid"

iuliana-prodan commented on src/include/sof/drivers/mu.h:
```diff
@@ -13,27 +13,61 @@
+
+#ifdef CONFIG_IMX8ULP
+/* Transmit Register */
+#define IMX_MU_xTRn(x)         (0x200 + 4 * (x))
```
Are these registers, `IMX_MU_xTRn`, `IMX_MU_xRRn`, used? 
I couldn't find them anywhere other than:
```
/sof$ grep -rnw IMX_MU_xTRn
tools/fuzzer/platform/mu.h:19:#define IMX_MU_xTRn(x)          (0x00 + 4 * (x))
src/include/sof/drivers/mu.h:17:#define IMX_MU_xTRn(x)		(0x00 + 4 * (x))
```

If they are not used, maybe we should remove them. Add them when need it.

iuliana-prodan commented on src/include/sof/drivers/mu.h:
```diff
@@ -13,27 +13,61 @@
-#define IMX_MU_xSR_RFn(x)	BIT(24 + (3 - (x)))
-#define IMX_MU_xSR_TEn(x)	BIT(20 + (3 - (x)))
+#define IMX_MU_xSR(type, index)		(type == IMX_MU_V2 ? index : 0x20)
+#define IMX_MU_xSR_GIPn(type, x)	(type == IMX_MU_V2 ? BIT(x) : BIT(28 + (3 - (x))))
```
Can you, please, align this?

iuliana-prodan commented on src/include/sof/drivers/mu.h:
```diff
@@ -13,27 +13,61 @@
-#define IMX_MU_xSR_GIPn(x)	BIT(28 + (3 - (x)))
-#define IMX_MU_xSR_RFn(x)	BIT(24 + (3 - (x)))
-#define IMX_MU_xSR_TEn(x)	BIT(20 + (3 - (x)))
+#define IMX_MU_xSR(type, index)		(type == IMX_MU_V2 ? index : 0x20)
```
Why do you have an `index` here? Shouldn't be a constant like 0x20?
Something like:
`#define IMX_MU_xSR(type, index) (type == IMX_MU_V2 ? const_val_for_version_2: 0x20)`

TangleZ replied:
It's not a constant, according to the code, it will be different register.

iuliana-prodan commented on src/include/sof/drivers/mu.h:
```diff
@@ -13,27 +13,61 @@
 
 /* Control Register */
-#define IMX_MU_xCR		0x24
+#define IMX_MU_xCR(type, index)		(type == IMX_MU_V2 ? index : 0x24)
```
Same as above -why do you need the `index`?
If this is not necessary you can also remove the `idx` argument from `imx_mu_xcr_rmw()`, `imx_mu_xsr_rmw()` functions.

TangleZ replied:
The same reason as before.

iuliana-prodan commented on src/include/sof/drivers/mu.h:
```diff
@@ -45,26 +79,26 @@ static inline void imx_mu_write(uint32_t val, uint32_t reg)
 }
 
-static inline uint32_t imx_mu_xcr_rmw(uint32_t set, uint32_t clr)
+static inline uint32_t imx_mu_xcr_rmw(int type, int idx, uint32_t set, uint32_t clr)
```
Please check if `idx` is really necessary.

TangleZ replied:
Yes. I'm sure this parameter is needed.

iuliana-prodan commented on src/drivers/imx/sai.c:
```diff
@@ -88,9 +89,16 @@ static void sai_start(struct dai *dai, int direction)
 			REG_SAI_MCTL_MCLK_EN);
 #endif
 
+	chan_idx = BIT(0);
```
What is `chan_idx`? 
Can you give us some context or maybe point us to the documentation.

iuliana-prodan commented on src/include/sof/drivers/sai.h:
```diff
@@ -18,6 +18,8 @@
 
 #ifdef CONFIG_IMX8M
 #define SAI_OFS		8
+#elif CONFIG_IMX8ULP
```
This can be written as:
```
#ifdef CONFIG_IMX8M || CONFIG_IMX8ULP
#define SAI_OFS		8
#else
#define SAI_OFS		0
#endif
```

iuliana-prodan commented on src/include/sof/drivers/sai.h:
```diff
@@ -251,8 +251,14 @@
 #define SAI_CLOCK_DIV		0x7
 #define SAI_TDM_SLOTS		2
+#else
+/* expect get frame clk is 16kHz, mono channel and 16bit word on 8ulp */
```
s/expect get frame clk is 16kHz, mono channel and 16bit word on 8ulp/ On 8ULP expected frame clk is 16kHz, mono channel and 16 bit word.

iuliana-prodan commented on src/include/sof/lib/dai.h:
```diff
@@ -417,6 +420,18 @@ static inline int dai_get_fifo(struct dai *dai, int direction,
+			       int stream_id)
+{
+	int ret = dai->drv->ops.get_srcid(dai, direction, stream_id);
+
```
Too many newlines.
Also, I believe you can return directly `dai->drv->ops.get_srcid(dai, direction, stream_id);`

iuliana-prodan commented on src/include/sof/platform.h:
```diff
@@ -47,6 +47,24 @@ int platform_init(struct sof *sof);
 
 /** @}*/
 
+#define HOST_TO_LOCAL 0
```
This is i.MX specific, even more is 8ULP specific.
Can we add this in a file imx specific?

TangleZ replied:
It should be common code, if we add new platform with need to convert address then this part code will be used. If we add it 
in 8ulp specific, and then we need make it be common when add new platform.

iuliana-prodan replied:
For now is only for 8ulp, so add it just for 8ulp, let's not change the core code if is not needed there.
IF, at some point, will be needed for something else we can make it common, but now this has nothing to do with other imx targets or Intel's.

TangleZ replied:
Sure.

iuliana-prodan commented on src/include/sof/lib/dma.h:
```diff
@@ -130,6 +130,7 @@ struct dma_sg_config {
 	uint32_t dest_dev;
 	uint32_t cyclic;			/* circular buffer */
 	uint64_t period;
+	int srcid;
```
Why do you add `srcid`? Where and for what do you use it?

I see is used in `edma.c` (line 338, above), but instead of 
```
if (srcid) {
...
} 
```
you can use an `#ifdef CONFIG_IMX8ULP`

TangleZ replied:
The reason is srcid may be a common resource, when edma really need. And this won't block existing platforms.

iuliana-prodan replied:
Commom resource for what? What is used for? 
All I understand right now is that `srcid` is for 8ULP.  But I don't know what it represent - a source id for what?
If is only used to determine it's 8ULP use instead of:
```
if (srcid) {
...
} 
```
you can use an `#ifdef CONFIG_IMX8ULP`

TangleZ replied:
The name srcid is a copy name from edma driver, I think its meaning should be a source id, it represent dmamux number.

iuliana-prodan commented on src/platform/imx8ulp/include/arch/xtensa/config/core-isa.h:
```diff
@@ -0,0 +1,673 @@
+#define XCHAL_NUM_EXTINTERRUPTS		27	/* num of external interrupts */
+#define XCHAL_NUM_INTLEVELS		4	/* number of interrupt levels
+						   (not including level zero) */
+#define XCHAL_EXCM_LEVEL		2	/* level masked by PS.EXCM */
```
Where did you take this values from?

TangleZ replied:
This file from ./crosstool-ng/.build/xtensa-imx8ulp-elf/build/overlay/newlib/newlib/libc/sys/xtensa/include/xtensa/config/core-isa.h

Commit:
drivers: imx: sai: Config channel and tx/rx for 8ULP

We expect mono pcm in imx8ulp platform, then config only one channel
works and enable tx0, rx2 for 8ULP.

Signed-off-by: Zhang Peng <peng.zhang_8@nxp.com>

Commit:
include: drivers: sai: Set SAI_FIFO_WORD_SIZE for imx8ulp

SAI has transmitter/receiver watermark is 16 on imx8ulp, then set
SAI_FIFO_WORD_SIZE 16.

Signed-off-by: Zhang Peng <peng.zhang_8@nxp.com>

iuliana-prodan commented on src/drivers/imx/edma.c:
```diff
@@ -324,6 +334,14 @@ static int edma_setup_tcd(struct dma_chan_data *channel, int16_t soff,
 		return rc;
 
+	/* only impact 8ulp platform */
+	if (srcid) {
```
IF only impacts 8ULP I would put this all in an `#ifdef CONFIG_IMX8ULP`, and remove the chan_addr_convert() function from above.
```
#ifdef CONFIG_IMX8ULP
        if (!dma_chan_reg_read(channel, EDMA_CH_MUX))
		dma_chan_reg_write(channel, EDMA_CH_MUX, srcid);
	if (channel->direction == DMA_DIR_MEM_TO_DEV)
		convert_addr(LOCAL_TO_HOST, sbase);
	else
		convert_addr(LOCAL_TO_HOST, dbase);
#endif
```
Is not ideal, but to add a variable, `srcid`, just for this I believe is worst.

Also, maybe we can add the `convert_addr()` in a` src/driver/imx/utils.c` file? 
@dbaluta @TangleZ what do you think?

dbaluta replied:
@TangleZ I agree with @iuliana-prodan. We could do this change locally to our imx edma driver and remove the changes in SOF core.
In the future if other drivers will need this we can reintroduce the get_srcid function.

So, as far as I understand srcid is to map between DMA events and DMA chnannels right? 
And this is done via a MUX?

Please explain this in the commit message and go with a solution that follows Iulia's suggestion which doesn't impact SOF core.

TangleZ replied:
@dbaluta @iuliana-prodan Not very clear about this. SRCID is not related to convert addr function, they can't be mixed.
If you think the code of srcid part is changed SOF core, then I can use hardcode to fix it in edma driver.
If you think convert addr part is changed SOF core, then we can do as iuliana-prodan said, create new file named utils.c for it.

iuliana-prodan commented on tools/topology/CMakeLists.txt:
```diff
@@ -136,6 +136,7 @@ set(TPLGS
 	"sof-jsl-da7219\;sof-jsl-da7219\;-DPLATFORM=jsl"
 	"sof-jsl-da7219\;sof-jsl-da7219-mx98360a\;-DPLATFORM=jsl-dedede"
 	"sof-imx8mp-wm8960\;sof-imx8mp-wm8960"
+	"sof-imx8ulp-btsco\;sof-imx8ulp-btsco"
```
Why did you added -btsco? 
Is this related with the Bluetooth SCO Audio - https://man.netbsd.org/btsco.4

TangleZ replied:
Yes. sai will connect to Bluetooth.

dbaluta commented on src/audio/dai.c:
```diff
@@ -838,6 +839,7 @@ static int dai_prepare(struct comp_dev *dev)
 		return -EINVAL;
 	}
+	dd->chan->srcid = dai_get_srcid(dd->dai, dai->direction, dd->stream_id);
 
```
We need to think about this. For non-8ulp platforms this function will crash. At least return 0 if src_id is not implemented.

dbaluta commented on src/include/sof/lib/dai.h:
```diff
@@ -417,6 +420,18 @@ static inline int dai_get_fifo(struct dai *dai, int direction,
+static inline int dai_get_srcid(struct dai *dai, int direction,
+			       int stream_id)
+{
+	int ret = dai->drv->ops.get_srcid(dai, direction, stream_id);
```
return 0 here if ops.get_srcid is NULL.

TangleZ replied:
Yes. It should be 0 if the platform is not 8ULP. Then edma won't set srcid.

iuliana-prodan commented on tools/topology/sof/pipe-volume-playback.m4:
```diff
@@ -80,7 +80,8 @@ indir(`define', concat(`PIPELINE_PCM_', PIPELINE_ID), Passthrough Playback PCM_I
 #
-PCM_CAPABILITIES(Passthrough Playback PCM_ID, CAPABILITY_FORMAT_NAME(PIPELINE_FORMAT), PCM_MIN_RATE, PCM_MAX_RATE, 2, PIPELINE_CHANNELS, 2, 16, 192, 16384, 65536, 65536)
+PCM_CAPABILITIES(Passthrough Playback PCM_ID, CAPABILITY_FORMAT_NAME(PIPELINE_FORMAT),
+			PCM_MIN_RATE, PCM_MAX_RATE, PIPELINE_CHANNELS, PIPELINE_CHANNELS, 2, 16, 192, 16384, 65536, 65536)
```
You removed 2 (as the channels_min) at all, you should change it to 1, not remove it at all.

TangleZ replied:
Not is removed it, just make channel_max = channel_min as in other files. Please refer pipe-volume-capture.m4
